core/log/solver_debug.cpp

```diff
@@ -0,0 +1,321 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <ginkgo/core/log/solver_debug.hpp>
+
+
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <string>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/mtx_io.hpp>
+#include <ginkgo/core/base/name_demangling.hpp>
+#include <ginkgo/core/log/logger.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+
+
+#include "core/base/dispatch_helper.hpp"
+
+
+namespace gko {
+namespace log {
+namespace {
+
+
+template <typename Functor>
+static bool dispatch_type(const LinOp* value, Functor fn)
```

Comment:
you could decouple the type checking from the dispatch, i.e. have
```
bool is_dense(const LinOp* value){
  return dynamic_cast<const ConvertibleTo<matrix::Dense<double>>*>(value) || 
         dynamic_cast<const ConvertibleTo<matrix::Dense<std::complex<double>>>*>(value);
}
```
And then use the run dispatch like
```
run<matrix::Dense, double, float, std::complex<float>, std::complex<double>>(
  value, [&](const auto* dense){
    auto host = make_temporary_clone(dense->get_executor()->get_master(), dense);
    stream << host->at(0, 0);
  });
```
I would at least prefer having the check by itself, because it would simplifiy the if construct in `print_scalar`.

Reply:
pinging this again. IMO the runtime check should at least be separated.

Reply:
going through `matrix::Dense` requires me to dispatch to all dense types, while going through ConvertibleTo only needs to choose between real/complex, that wouldn't really simplify the code. Do you mean using `run<ConvertibleTo, matrix::Dense<double>, matrix::Dense<float>>`?

Reply:
My main point is how the dispatch is used. I'm not a fan of the
```c++
} else {
    if (!dispatch_type(
            value, [&](auto vector) { stream << vector->at(0, 0); })) {
        stream << "<unknown>";
    }
}
```
construct.
By separating the runtime check you would have
```c++
} else if(is_dense(vector)) {
    // dispatch how you want
} else {
    stream << "<unknown>";
}
```
which I find more digestible.

Reply:
the code became a bit more complex due to the need for `detail::pointee<decltype(value)>::result_type`, but I think it's still a clearer separation, so I applied the suggestion.